File: Trees/LC_98.cpp
// Written by: Jian-Yuan Lin
// Date: 2022.11.29

class Solution {
public:
    bool isValidBST(TreeNode* root) {
        // in CPP, the minval and maxval we need to use long
        // since it's possible that the root value is INT_MAX
        return check(root, LLONG_MIN, LLONG_MAX);
    }
private:
    bool check(TreeNode* root, long minval, long maxval){
        if (root == nullptr){
            return true;
        }
        if (root->val <= minval || root->val >= maxval){
            return false;
        }
        return check(root->left, minval, root->val) && check(root->right, root->val, maxval);
    }
};

File: Trees/LC_98.py
# Written by: Jian-Yuan Lin
# Date: 2022.11.29

class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        def check(root, minval, maxval):
            if not root:
                return True
            if root.val <= minval or root.val >= maxval:
                return False
            return check(root.left, minval, root.val) and check(root.right, root.val, maxval)
        return check(root, -float('inf'), float('inf'))